modules/lib/core/index.d.ts

```diff
@@ -118,3 +118,472 @@ export interface Resource {
+    query: string;
+    operator?: DslOperator;
+    boost?: number;
+}
+
+export interface NgramDslExpression {
+    fields: string[];
+    query: string;
+    operator?: DslOperator;
+    boost?: number;
+}
+
+export interface StemmedDslExpression {
+    fields: string[];
+    query: string;
+    language: string;
+    operator?: DslOperator;
+    boost?: number;
+}
+
+export interface ExistsDslExpression {
+    field: string;
+    boost?: number;
+}
+
+export interface BooleanDslExpression {
+    should?: QueryDsl | QueryDsl[];
+    must?: QueryDsl | QueryDsl[];
+    mustNot?: QueryDsl | QueryDsl[];
+    filter?: QueryDsl | QueryDsl[];
+    boost?: number;
+}
+
+export type QueryDsl = {
+    boolean: BooleanDslExpression;
+} | {
+    ngram: NgramDslExpression;
+} | {
+    stemmed: StemmedDslExpression;
+} | {
+    fulltext: FulltextDslExpression;
+} | {
+    matchAll: MatchAllDslExpression;
+} | {
+    pathMatch: PathMatchDslExpression;
+} | {
+    range: RangeDslExpression;
+} | {
+    like: LikeDslExpression;
+} | {
+    in: InDslExpression;
+} | {
+    term: TermDslExpression;
+} | {
+    exists: ExistsDslExpression;
+};
+
+type SortDirection = 'ASC' | 'DESC';
+
+type DistanceUnit =
+    | 'm'
+    | 'meters'
+    | 'in'
+    | 'inch'
+    | 'yd'
+    | 'yards'
+    | 'ft'
+    | 'feet'
+    | 'km'
+    | 'kilometers'
+    | 'NM'
+    | 'nmi'
+    | 'nauticalmiles'
+    | 'mm'
+    | 'millimeters'
+    | 'cm'
+    | 'centimeters'
+    | 'mi'
+    | 'miles';
+
+export interface FieldSortDsl {
+    field: string;
+    direction?: SortDirection;
+}
+
+export interface GeoDistanceSortDsl
+    extends FieldSortDsl {
+
+    unit?: DistanceUnit;
+    location?: {
+        lat: number;
+        lon: number;
+    }
+}
+
+export type SortDsl = FieldSortDsl | GeoDistanceSortDsl;
+
+//
+// START AGGREGATIONS, FILTERS
+//
+export interface Bucket {
+    [subAggregationName: string]: AggregationsResult | string | number | undefined;
+
+    key: string;
+    docCount: number;
+}
+
+export interface NumericBucket
+    extends Bucket {
+    from?: number;
+    to?: number;
+}
+
+export interface DateBucket
+    extends Bucket {
+    from?: string;
+    to?: string;
+}
+
+export interface BucketsAggregationResult {
+    buckets: (DateBucket | NumericBucket)[];
+}
+
+export interface StatsAggregationResult {
+    count: number;
+    min: number;
+    max: number;
+    avg: number;
+    sum: number;
+}
+
+export interface SingleValueMetricAggregationResult {
+    value: number;
+}
+
+export type AggregationsResult = BucketsAggregationResult | StatsAggregationResult | SingleValueMetricAggregationResult;
+
+export type Aggregation =
+    | TermsAggregation
+    | HistogramAggregation
+    | DateHistogramAggregation
+    | NumericRangeAggregation
+    | DateRangeAggregation
+    | StatsAggregation
+    | GeoDistanceAggregation
+    | MinAggregation
+    | MaxAggregation
+    | ValueCountAggregation;
+
+export interface TermsAggregation {
+    terms: {
```

Comment:
It would be nice if this was "split" up even more

```
export interface TermsAggregationParams {
        field: string;
        order?: string;
        size?: number;
        minDocCount?: number;
}
    
export interface TermsAggregation {
    terms: TermsAggregationParams;
    aggregations?: Record<string, Aggregation>;
}
```

Because that's certainly something I've used. The workaround for me is using TermsAggregation['terms'], but that's not very pretty

Reply:
I actually find `TermsAggregation['terms']` to be preferable to extracting it out. 

There are so many interfaces already, and this helps keeping the "sub type" in the context it is used in.

@ComLock In the end, I think it might be more about what you eyes are used to. Personally I'm happy when I see types like `TermsAggregation['terms']`, because it also tells me about where this value is intended to end up in a larger context.

Reply:
Yeah, on one side it's cleaner to use `TermsAggregationParams`, on the other side the context is less clear and the imports/exports become dirtier. 
What I usually do in my projects in similar cases is `type TermsAggregationParams = TermsAggregation['terms']`. But it is only okay if you need to do this in 1-2 files in a whole project. 
Are developers (and you guys) are using `TermsAggregation['terms']` often? If it's a pretty rare case, then I suggest we keep it as is and just "rename" it locally in your project.